### PR TITLE
34 create gtfs data extraction/ingestion script for all three cities transit agencies

### PR DESCRIPTION
**Pull request by mbjackson-capp**
Main functionality: Extract the static GTFS transit data for all three of our pilot cities, add columns to give each asset a unique ID, and combine the different agencies' GTFS feeds into tall multi-city dataframes to be put into Postgres backend. Linked issue: #34 

Note: Because, as discussed at Monday's post-class standup, PostGIS install is having some issues, I have not yet implemented functions that convert lat/long columns into GIS Point() objects, and we may not be able to feed in the LineString() information directly. Also, it is already known that the built-in `geometrize_routes()` function in the gtfs-kit library, which turns the long list of points in a GTFS feed's `shapes.txt` into mappable LineString() objects, does not work for some transit systems (the error is pretty deep in the stack trace and I haven't figured it out yet.) **We should open another issue for creating functions that reliably, correctly convert from numeric data to Point() and LineString() objects.** 
- If we don't get working PostGIS, this conversion should happen AFTER the user queries data from the backend; if we do, the conversion should be added into this file and be done BEFORE the data goes into the backend.

Screenshot of tests passing:
Screen Shot 2024-04-24 at 10 41 30 AM

### CODE REVIEW

File: backend/tests/test_extract_scheduled_gtfs.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    # TODO: consider making this a fixture
+    chicago_combined_dfs = combine_different_feeds([CTA_URL, METRA_URL])
+    assert "routes" in chicago_combined_dfs
+    assert "shapes" in chicago_combined_dfs
+    assert "trips" in chicago_combined_dfs
+    assert "stops" in chicago_combined_dfs
+    assert "stop_times" in chicago_combined_dfs
+    assert "transfers" in chicago_combined_dfs
+    # TODO: more detailed testing to make sure "Chicago_CTA_" and "Chicago_Metra_"
+    # are represented in the uniq_id column for each sub-sheet (for now, that
```
Comment (meganhmoore) [inline backend/tests/test_extract_scheduled_gtfs.py:89]: great tests!

File: backend/ingestion/extract_scheduled_gtfs.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+
+def ingest_gtfs_feed(feed_url: str) -> tuple:
+    """
+    Use gtfs_kit library to ingest the most up-to-date scheduled (static) GTFS
+    data from a particular URL, and return each component file as a GeoDataFrame.
+    """
+    url_base = re.findall(r"https?:\/\/[^\/]+\/", feed_url)[0]
+    feed_city = URL_TO_CITY[url_base]["City"]
+    feed_agency = URL_TO_CITY[url_base]["Agency"]
+    feed = gk.read_feed(feed_url, dist_units="km")  # dist_units TBD
```
Comment (meganhmoore) [inline backend/ingestion/extract_scheduled_gtfs.py:99]: just out of curiosity, is it possible to get miles instead? My sense is that will be a more standard unit for US transit but can't see any problems with it for now

File: backend/ingestion/extract_scheduled_gtfs.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+    # TODO: miscellaneous cleanup (one thing: standardize CTA route names to
+    # standard English words for colors)
+    # TODO: Consider adding a # sign to color column (or see if Postgres has a
+    # color data type)
+    return feed_component
+
+
+def combine_different_feeds(
+    feed_url_list: list[str], output_folder: None | str = None
+) -> dict[pd.DataFrame | gpd.GeoDataFrame]:
```
Comment (meganhmoore) [inline backend/ingestion/extract_scheduled_gtfs.py:224]: nice use of types!!

File: backend/ingestion/extract_scheduled_gtfs.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+    print(all_dfs)
+
+    if output_folder is not None:
+        for name, df in all_dfs.items():
+            pd.to_csv(df, f"{output_folder}/{name}.csv")
+
+    return all_dfs
+
+
+# TODO: function to convert lat/long and shapes.txt data into geometry objects.
```
Comment (meganhmoore) [inline backend/ingestion/extract_scheduled_gtfs.py:290]: 👍 can help w using the [shapely](https://pypi.org/project/shapely/) tools if you end up going this route

**Comment by meganhmoore** [comment]
approved from me once merge conflicts are resolved! lmk if you need help with them.

**Review by meganhmoore — APPROVED** [review]

**Review by jimenasalinas — APPROVED** [review]
Great use of comments, it helped me follow what was going on in the code. I appreciate how your functions are clean and incorporate the best practices that we're learning in class. 

I noticed this comment: "TODO: Question -- do we want to create a new primary key that concatenates
    # city and agency information, or just do JOINs that add city and agency
    # to the ON condition to make sure all JOINs are within a city-agency system?". We should check-in with JP about this before the ingestion! My assumption is that we would do City-id at this point so we can each just proceed to ingesting our data into the db.

File: backend/ingestion/extract_scheduled_gtfs.py (hunk 2 above)
Comment (mbjackson-capp) [reply backend/ingestion/extract_scheduled_gtfs.py:99]: yep - can just change `dist_units="km"` to `dist_units="mi"`

**Pull request merged**